torch/utils/checkpoint.py

```diff
@@ -0,0 +1,118 @@
+import torch
+from torch.autograd import Variable, Function
+
+
+def repackage_inputs(inputs):
+    if torch.is_tensor(inputs):
+        return Variable(inputs)
+    elif isinstance(inputs, tuple):
+        return tuple(repackage_inputs(v) for v in inputs)
+    else:
+        raise RuntimeError("Unknown input type")
+
+
+def unpack_variables(inputs):
+    if type(inputs) == Variable:
+        return inputs.data
+    elif torch.is_tensor(inputs):
+        return inputs
+    elif isinstance(inputs, tuple):
+        return tuple(unpack_variables(v) for v in inputs)
+    else:
+        raise RuntimeError("Unknown input type")
+
+
+class CheckpointFunction(Function):
+
+    # NOTE: *args is the flat inputs list, args is the tuple containing inputs
+    @staticmethod
+    def forward(ctx, run_function, *args):
+        ctx.run_function = run_function
+        ctx.save_for_backward(*args)
+        inputs = repackage_inputs(args)
+        with torch.no_grad():
+            outputs = run_function(*inputs)     # the *inputs* is always a tuple
+        return unpack_variables(outputs)
+
+    @staticmethod
+    def backward(ctx, *args):
+        saved_inputs = ctx.saved_tensors
+        inputs = repackage_inputs(saved_inputs)
```

Comment:
The problem with in-place checks is caused exactly by this repackaging, so it's not fixed yet. BTW does it work now? It seems like you're repackaging inputs but not setting their `requires_grad`. The way to solve this would be to use `ctx.saved_variables`, and then `torch.autograd.grad(output, grad_otuputs, saved_variables)`

Reply:
yes, it still works, I tested all the models again and also the unit test here. 

I had tried the ```saved_variables``` and ```torch.autograd.grad``` suggestion as well but it throws error ```RuntimeError: One of the differentiated Variables does not require grad```

Reply:
Hmm yeah that's because `torch.autograd.grad` requires you to only give Variables that do require grad as its third argument. One way to fix it would be to filter out those that don't, and then reconstruct a list of `grad_inputs` by inserting `None`s in correct places. We won't get correctness checks without it

Reply:
alright, chatted with offline and there is known bug in autograd that stops from using the ```torch.autograd.grad``` for now. Adam is working on fixing it with ETA Wednesday!